New guide for making addon

wangyikewxgm opened the pull request:
1. new guide for making an addon
2. self signed helm repo and OCI registry


Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

### Description of your changes

<!--

Briefly describe what this pull request does. We love pull requests that resolve an open KubeVela issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] Update `sidebar.js` if adding a new page.
- [ ] Run `yarn start` to ensure the changes has taken effect.


### Special notes for your reviewer

<!--

Be sure to direct your reviewers'
attention to anything that needs special consideration.

-->

wangyikewxgm committed:
in the middle of something

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

finish addon

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

finish helm and adon

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

fix

fix annotation

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

finish update

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
solve conflict

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow reviewed (commented):
make this picture larger and more clear: docs/resources/addon-mechanism.jpg

wonderflow reviewed (commented):
先写中文

wangyikewxgm committed:
update addon docs and update english

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
fluxcd addon only helm components

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/reference/addons/fluxcd.md:
````diff
@@ -12,6 +12,12 @@ title: FluxCD
 vela addon enable fluxcd
 ```
 
+如果你只想利用 fluxcd addon 部署 helm 应用，你就可以执行下面的命令启只启用 helm 相关的组件：
````
```suggestion
如果你只想利用 fluxcd addon 部署 helm 应用，你可以执行下面的命令启只启用 helm 相关的组件：
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,42 @@
 
-一个 KubeVela 的插件就是包含了一系列文件的目录。 下图展示了在启用一个插件时，KubeVela 做了哪些事情, 主要包含以下三个步骤。
+一个 KubeVela 插件就是一个可以包含了以下两类文件的集合:
+* 定义 Kubernetes 资源对象的`资源描述文件`。他们可以是一个支撑某个模版定义文件背后的 kubernetes operator，或者是一个 webserver 等。
```
这里提 webserver 代表什么？  Kubernetes operator是有特指的可以理解。

wangyikewxgm replied:
改成了：其他的 kubernetes 应用

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,42 @@
-* 接下来运行在管控集群的 KubeVela 控制器根据需要在管控平面或者子集群中下发资源。
+下图展示了在启用一个插件时，KubeVela 做了哪些事情, 主要包含以下三个步骤。
+* 当通过 UX/CLI 启用一个插件时，会从插件仓库把插件中的源文件拉取下来。
+* 文件中`扩展配置`文件会被 UX/CLI 直接下发到管控集群。 资源描述（ resources, template 和 metadata ）文件会被用来渲染成一个 KubeVela 应用并创建。
```
resources, template 和 metadata 这些可以加个小链接链到下面的部分。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,42 @@
+下图展示了在启用一个插件时，KubeVela 做了哪些事情, 主要包含以下三个步骤。
+* 当通过 UX/CLI 启用一个插件时，会从插件仓库把插件中的源文件拉取下来。
+* 文件中`扩展配置`文件会被 UX/CLI 直接下发到管控集群。 资源描述（ resources, template 和 metadata ）文件会被用来渲染成一个 KubeVela 应用并创建。
+* 接下来运行在管控集群的 KubeVela 控制器根据需要在管控平面或者子集群中下发 kubernetes 资源。
```
```suggestion
* 接下来运行在管控集群的 KubeVela 控制器完全按照应用（Application）的执行流程交付资源，与 KubeVela 其他的应用执行没有差别。
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -72,12 +72,10 @@ invisible: false
 | system.vela     | no  | string | 环境中所要求的 KubeVela 的版本，如果不满足，安装将会被拒绝  |
 | system.kubernetes     | no  | string | 环境中所要求的 Kubernetes 的版本 |
-| needNamespace     | no  | []string | 安装之前需要创建的 namespace ， kubeVela 在插件安装开始之前会在所有集群上创建该 namespace   |
-| invisible     | no  | bool | 是否不可见，当插件尚在未完成阶段，可以设置暂时不可被启用 |
```
这个是删了吗

wangyikewxgm replied:
代码层面还没有删，但这两个特性今天其实都没用了：
1. invisible 主要是由于之前 addon 没有多版本，线上发现了一个bug，设置一个invisible 暂时让用户不可见，但现在有了多版本，这个意义不大了。
2. needNamespace 的作用是创建一个namesapce 放在第一个component，保证先下发它，但是workflow改动之后 component 不是按照顺序下发的了。所以这个字段意义也不大了。
所以不建议用户再了解这两项配置了。

wonderflow replied:
ok

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -88,138 +86,419 @@ invisible: false
-[实验阶段的插件](https://github.com/kubevela/catalog/tree/master/experimental/addons)通常没有这些非常严格的规则，但对于一个想要进阶到 [认证仓库](https://github.com/kubevela/catalog/tree/master/addons) 的插件而言，README 至关重要。
+[实验阶段的插件](https://github.com/kubevela/catalog/tree/master/experimental/addons) 通常没有这些非常严格的规则，但对于一个想要进阶到 [认证仓库](https://github.com/kubevela/catalog/tree/master/addons) 的插件而言，README 至关重要。
+
+### 应用模版 (template.yaml 或 template.cue) 文件 (非必须)
```
```suggestion
### 应用模版 (template.yaml 或 template.cue) 文件 (可缺省)
```

wonderflow replied:
我感觉混在一起不好，不如单独起个段落，叫完全使用 CUE 定义 addon。 老的yaml 模式就保留。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -88,138 +86,419 @@ invisible: false
+
+### 应用模版 (template.yaml 或 template.cue) 文件 (非必须)
+
+通过上面的介绍，我们知道插件中很重要的一部分就是通过`资源描述文件`定义需要安装的 kubernetes 资源对象。你可以在应用模版文件 (template.yaml 或 template.cue) 和 `resources/` 目录下的文件中定义这些资源。而这些资源描述文件会最终被渲染成一个 KubeVela 的 `应用`，由 KubeVela 控制器下发到各个集群。 所以这里的应用模板文件就是用来定义这个`应用`的基础框架。你可以通过该文件描述这个`应用`的特定信息，比如你可以为`应用`打上特定的标签或注解，当然你也可以直接在该应用模版文件中添加组件，策略和设置工作流。 应用模版定义文件可以是 YAML 类型的文件（template.yaml）或 CUE 类型文件 （template.cue）。
```
Kubernetes  首字母大写

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -88,138 +86,419 @@ invisible: false
+通过上面的介绍，我们知道插件中很重要的一部分就是通过`资源描述文件`定义需要安装的 kubernetes 资源对象。你可以在应用模版文件 (template.yaml 或 template.cue) 和 `resources/` 目录下的文件中定义这些资源。而这些资源描述文件会最终被渲染成一个 KubeVela 的 `应用`，由 KubeVela 控制器下发到各个集群。 所以这里的应用模板文件就是用来定义这个`应用`的基础框架。你可以通过该文件描述这个`应用`的特定信息，比如你可以为`应用`打上特定的标签或注解，当然你也可以直接在该应用模版文件中添加组件，策略和设置工作流。 应用模版定义文件可以是 YAML 类型的文件（template.yaml）或 CUE 类型文件 （template.cue）。
 
-### 应用模版 (template.yaml) 文件 (非必须)
+选择其中的哪个取决于你是否期望在启用插件时，通过启用参数动态变更应用当中的内容。
```
加一个小节，如何选择？ 几句话介绍一下利弊，功能差异。

大体上表达： yaml 就是简单直接功能少，适合全部用 yaml 只是简单部署的场景。 CUE 则功能强，可以多种渲染、部分部署等。

wonderflow replied:
分开以后这里就不需要对比了，直接介绍 YAML 就行，然后可以提一些不支持的功能，引用到 下面用 CUE 制作 addon  的文档。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -88,138 +86,419 @@ invisible: false
 
+#### template.yaml
+
+如果你不希望在启用时动态的变更资源的某些字段，`template.yaml` 就是足够的了。下面就是一个例子：
```
类似这种，放到上面去方便对比。 这里开头应该描述 `template.yaml` 本质上就是 KubeVela Application。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -88,138 +86,419 @@ invisible: false
+* [OCM control plane](https://github.com/kubevela/catalog/blob/master/addons/ocm-hub-control-plane/template.yaml) 在这个例子中使用的是 YAML 模版文件，应用应用中的资源只需要在管控集群中创建，且不需要在启用时更改字段。
+* [ingress-nginx](https://github.com/kubevela/catalog/tree/master/addons/ingress-nginx/template.cue) 这个例子中使用 CUE 模版定义文件。
+
+### parameter.cue （非必须）
```
这个应该放到 template.cue 里面讲吧？

wangyikewxgm committed:
split cue related out

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

fix error link issue

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,46 @@
 ![alt](../../resources/addon-mechanism.jpg)
 
-## 制作插件 (Make an Addon)
+# 制作插件 (Make an Addon)
```
这个一级目录效果看看是怎么样的，之前看效果展示好像有问题。

wangyikewxgm replied:
确实侧边栏会展示不出来，改回到二级目录了

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,46 @@
 ├── metadata.yaml
-└── template.yaml
+├── parameter.cue
+└── template.yaml(or template.cue)
```
这里yaml 和 cue 分开。可以只写yaml。然后加一句话解释都可以写成CUE

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
````diff
@@ -2,39 +2,46 @@
 ```
 
-需要注意的是，上面的文件并不都是必须的。接下来将介绍该目录下的每个资源文件和子目录的详细作用。
+你也可以使用命令 `vela addon init`（vela CLI v1.5 或更新）帮你快速创建这个目录框架，你可以查阅 `vela addon init -h`来获取详细信息。这里我们使用最简单的 `vela addon init your-addon-name` 就足够了。
````
1.5 已经发了，不需要特殊说明

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -72,211 +76,144 @@ invisible: false
+需要注意的是，应用模版文件和资源目录下面的资源描述文件，都可以通过 CUE 或 YAML 格式定义，本文档会只介绍最简单的通过 YAML 描述资源的方法，[下一篇文档](./addon-cue) 会详细介绍如何如果通过 CUE 定义资源描述文件。
 
-如果你需要为应用添加一个需要在启用时根据参数动态渲染的组件，你就可以编写一个 CUE 格式的文件，如下所示。
+### 应用模版 (template.yaml 或 template.cue) 文件
```
只说明 yaml 即可

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,435 @@
+---
+title: CUE 定义资源描述文件
```
```suggestion
title: 使用 CUE 制作自定义插件
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,435 @@
+title: CUE 定义资源描述文件
+---
+
+本文档将会详细介绍如何通过 CUE 编写应用模版文件 (template.cue) 和 `resources/` 目录下的资源文件来定义需要下发的资源。
```
先提一下使用 YAML 的那个文档。讲一下区别。

只讲 YAML 的也要带一下这个。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,435 @@
+
+* 资源可以在启用插件时通过启动参数动态渲染结果。
+* 插件中某些特殊的资源并不想作为应用的一部分在多集群范围内下发，而只想仅在管控中被创建。
+* 插件中可能包含多个 X-definitions 和背后支撑这些能力的 operator，他们能够根据启动参数被动态的被启用。
```
额外的功能还包括：

* 灵活的使用 CUE 语法定义、部署额外的资源

barnettZQG commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,3 @@
+---
+title: Make addon with cue
+---
```
Set to draft.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -2,16 +2,20 @@
 
+A KubeVela addon is a collection that can contain the following two types of files；
+* `Extension configuration file` that defines KubeVela extensibility points, such as [X-definitions](../../getting-started/definition), [UI-Schema](../../reference/ui-schema) or [topology-rules](../../reference/topology-rule).
+* `Resource description file` that defines Kubernetes resource objects. They may be the kubernetes operator behind a X-definition, or the webserver, etc.
```
kubernetes -> Kubernetes

barnettZQG reviewed (approved):
LGTM

wangyikewxgm committed:
fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,4 @@
+---
+title: Make addon with cue
+draft: true
```
why it's draft, we need finish before merge

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,4 @@
+---
+title: Make addon with cue
```
```suggestion
title: Build Addon with CUE
```

wonderflow commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -2,16 +2,20 @@
+
 The picture below shows what KubeVela does when an addon is enabled. There are mainly three process:
-* [Addon Registry](./addon-registry) store addons which can be used to share and distribute addons anywhere, it can be any git repo, OSS bucket or OCI registry.
+* [Addon Registry](./addon-registry) store addons which can be used to share and distribute addons anywhere, it can be any git repo, helm repo.
```
```suggestion
* [Addon Registry](./addon-registry) store addons which can be used to share and distribute addons anywhere, it can be any git or helm chart repository.
```

wonderflow commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -2,16 +2,20 @@
 ![alt](../../resources/addon-mechanism.jpg)
 
-## Build an addon
+## Make an addon
```
为什么是 Make 而不是 Build？

wangyikewxgm replied:
改成 make 一是表达“制作”的意义，二是避免 build 给开发者带来似乎是在构建一个二进制的错觉

wonderflow replied:
那你得整体改啊，这半吊子。。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+通过编写 CUE 的资源文件，插件将会获得以下能力：
+
+* 资源可以在启用插件时通过启动参数动态渲染结果。
+* 利用 CUE 语言灵活的语法和内置丰富的功能渲染和部署附属资源。
```
```suggestion
* 利用 CUE 语言灵活简洁的语法、丰富的内置函数及其参数校验能力，渲染和部署资源。
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+
+通过编写 CUE 的资源文件，插件将会获得以下能力：
+
+* 资源可以在启用插件时通过启动参数动态渲染结果。
```
```suggestion
* 资源可以在启用插件时通过启动参数动态渲染。
```

这一条和第三条重了吧，说的一回事

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+
+* 资源可以在启用插件时通过启动参数动态渲染结果。
+* 利用 CUE 语言灵活的语法和内置丰富的功能渲染和部署附属资源。
+* 插件中可能包含多个 X-definitions 和背后支撑这些能力的 operator，他们能够根据启动参数被动态的被启用。
```
现在不说 X-Definition 了，直接说 Definition

```suggestion
* 插件中可能包含多个模块定义（Definition）文件及其背后支撑的 CRD Operator，他们能够根据启动参数被选择性安装。
```

什么叫“被动态的启用”？

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+* 利用 CUE 语言灵活的语法和内置丰富的功能渲染和部署附属资源。
+* 插件中可能包含多个 X-definitions 和背后支撑这些能力的 operator，他们能够根据启动参数被动态的被启用。
+
+通过 CUE 定义插件和通过 YAML 定义插件的主要区别是在 `资源描述文件`（template.cue 和 resources 目录下的资源描述文件）。而`基本信息文件`和`扩展配置文件`定义的方式和 YAML 定义方式完全一直，具体请参考[文档](./intro)。
```
* 一直？
* 主要区别不是语言层面吗？ 你这里说的涉及文件范围的区别。
* 什么情况下的具体信息去参考那个文档，上面已经提到“这个参考”一次了。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+
+## 应用模版文件（template.cue）
+
+CUE 格式的应用模版描述文件中最重要的是定义一个 `output` 的 CUE 代码块，这个代码块所包含的必须是一个 KubeVela 的`应用`（application）。如下所示：
```
这里应该给一个引用，什么是 KubeVela Application？

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,438 @@
+$ vela addon enable <addon-name> namespace=my-namespace
+```
+
+经过渲染之后，`应用`如下所示：
````
```suggestion
不指定 namespace 的情况下，经过渲染的`应用`如下所示：
```

wangyikewxgm replied:
上面的命令是指定namespace的吧

wonderflow replied:
你下面的例子不是 `vela-system`吗？ 那你要一致啊

wangyikewxgm replied:
addon application 自身的namespace 是 vela-system 这个没有被参数所设置，这个例子中设置的是应用中 k8s-objects 组件中创建的 namespace 资源，他是被参数设置为了 `my-namespace`

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,438 @@
+              name: my-namespace
+```
+
+> 需要注意的是，即使你通过 `metadata.name` 字段设置了应用的名称，该设置也不会生效，在启用时应用会统一以 addon-{addonName} 的格式被命名。
````
```suggestion
> 需要注意的是，即使你通过 `metadata.name` 字段设置了应用的名称，该设置也不会生效，在启用时应用会统一以 addon-{addonName} 的格式自动命名。
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+
+### 如何定义 `template.cue` 实现根据参数选择安装集群
+
+如果你希望插件中的资源能够不止被安装在管控集群，同时也能够安装在各个其他的子集群中， 并且具体安装在哪些集群需要由用户在启用插件时设置 `clusters` 启动参数来指定。 你就可以在 `template.cue` 中添加一个 topology 策略来实现，如下所示：
```
topology 是什么？

wonderflow replied:
逻辑上应该先解释“这个是 KubeVela Application 的功能，两个严格一致”，然后再展开 Addon 里如果要做这些功能，可以利用 Application 的这些能力。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,438 @@
+_rules: {...}
+```
+
+这个例子中我们定义了一个 `resourceTree` 的 configmap ，这个 configmap 其实是一个[拓扑树资源关系规则](../../reference/topology-rule)，这个 configmap 就只需要在管控集群中部署。
````
资源拓扑的能力是个很强的功能，不应该这样一笔带过。

wonderflow replied:
而且不只是看到拓扑这么简单，还能使用 vela exec、port-forward 之类的命令行能力。

wangyikewxgm replied:
考虑之后提个单独的pr完善 topology-rule 的介绍文档，这个地方只作为一个 附属资源的例子。

wonderflow replied:
这里还是应该讲一下好处，一句话就足够。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,438 @@
+          - cluster1
+```
+
+> 需要注意的是，CUE 资源描述文件需要必须要包含一个 `package main` header 才会被识别为一个资源描述文件，从而与应用模版文件放在一个渲染上下文中渲染。 
````
换个视角，你要解释这么写（`package main`）的好处而不是告诉用户你必须要背负这个“负担”。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,438 @@
+
+### 使用 context 中的变量渲染组件
+
+除了通过启动参数动态渲染应用外，你还可以读取 `metadata.yaml` 中定义的字段来做渲染。 例如你可以如下定义一个 `templatec.cue` 文件：
```
templatec.cue 又是什么？

wonderflow reviewed (changes requested):
still many issues on this doc

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,46 @@
 
-一个 KubeVela 的插件就是包含了一系列文件的目录。 下图展示了在启用一个插件时，KubeVela 做了哪些事情, 主要包含以下三个步骤。
+一个 KubeVela 插件就是一个主要包含了以下三类文件的集合:
+* 插件的`基本信息文件`，这类文件是制作一个插件所必须具备的文件。包括元数据文件（metadata.yaml）和 自描述文件（README.md）。
```
```suggestion
* 插件的`基本信息文件`，这类文件是制作一个插件所必须具备的文件。包括元数据文件（metadata.yaml）和插件的介绍文档（README.md）。
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -72,204 +76,138 @@ invisible: false
+| deployTo.runtimeCluster     | no  | bool |  插件是否可以安装到子集群，默认不设置该字段插件不会安装在任何子集群中 |
 
-### 自描述  (README.md) 文件 (必须)
+#### 自描述  (README.md) 文件 
```
```suggestion
#### 插件文档  (README.md) 
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -72,204 +76,138 @@ invisible: false
-### 自描述  (README.md) 文件 (必须)
+#### 自描述  (README.md) 文件 
 
 插件的自描述文件用于描述插件的主要功能，并且该文件会在 UX 的插件详情页面呈现给用户。所以 README.md 需要包含以下基本信息：
```
整体把自描述这个词改了，感觉是机器翻译的

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -72,204 +76,138 @@ invisible: false
 * 为什么要使用该插件？使用案例或场景
 * 如何使用？ `End user` 能够快速理解如何使用该插件。最好能够提供一个端到端的例子。
-* 安装了什么？ 插件中包含的各个 `definition` 以及相关的 CRD 和 controller。
+* 安装了什么？ 插件中包含的 `X-definition` 以及相关的 CRD 和 controller。
```
没有 X-Definition 这个说法，我们都用这个说法：

模块化能力（Definition）

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
````diff
@@ -72,204 +76,138 @@ invisible: false
-	}]
-}
-```
+应用模板文件就是用来定义这个`应用` （application） 的基础框架。你可以通过该文件描述这个`应用`的特定信息，比如你可以为`应用`打上特定的标签或注解，当然你也可以直接在该应用模版文件中添加组件，策略和设置工作流。 应用模版定义文件可以是 YAML 类型的文件（template.yaml）或 CUE 类型文件 （template.cue）。
````
最好有个地方先解释一下既可以直接用YAML，也可以用CUE，这里就不用混着介绍了。

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -89,9 +88,17 @@ The README of this addon, it will be displayed in the dashboard for end user who
-### template.yaml(Optional)
+### template file (Optional)
+
+Through the above introduction, we know that a very important part of an addon is resource description file that define the Kubernetes resource objects that needs to be installed. You can write the `template file` (template.yaml or template.cue) or files below `resources/` dir to define these resources. These resources will be rendered into a KubeVela application, which applied to each cluster by ...
```
Through the above introduction, we know that a very important part of an addon is its resource description files which define the Kubernetes resource objects to be installed.

Somefive replied:
which will be dispatched to clusters by the KubeVela controller once the addon is enabled.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -89,9 +88,17 @@ The README of this addon, it will be displayed in the dashboard for end user who
+
+Through the above introduction, we know that a very important part of an addon is resource description file that define the Kubernetes resource objects that needs to be installed. You can write the `template file` (template.yaml or template.cue) or files below `resources/` dir to define these resources. These resources will be rendered into a KubeVela application, which applied to each cluster by ...
+
+Which to choose depends on whether you want to change the application by parameters when the addon is enabled.
```
Depending on the need of parameterizing the application when enabling the addon, you can select the preferred way to customize your design.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -89,9 +88,17 @@ The README of this addon, it will be displayed in the dashboard for end user who
+
+Which to choose depends on whether you want to change the application by parameters when the addon is enabled.
+
+>Please notice: You can only choose one of `template.cue` or `template.yaml`, otherwise will report an error when enable the addon.
```
otherwise an error will be reported when enabling the addon.

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
 ```
 
-You can describe anything in the application if you want, the rest resources defined in other folders will be automatically rendered and appended into this template as components.
+In this example, we define the basic framework of an application. This application contains a `k8s-objects` type component that contains `namespace` resource. After the addon is enabled, this namespace will be applied by KubeVela to all clusters.
````
we define the skeleton of an application.

Somefive replied:
`k8s-objects` typed component that contains one `namespace` typed resource.

Somefive replied:
will be applied to all clusters by the KubeVela controller.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+#### template.cue
 
-> If the addon will be installed in managed clusters, please don't define policies or workflow steps in the template, or it will break the automatic precess built internally for addon installation.
+If you want the end user to determine some fields of the application by setting parameters when enabling the addon, you need write a `template.cue` file to do so. You can define the `template.cue` file like this to allow end-user specify the name of namespace in the above example:
```
like this -> like the following code

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
+}
+```
+
+As we can see, this CUE file mainly contain a `ouput` block. The `output` block must be a KubeVela application that may contain components, policies, or workflow. 
````
The CUE file shown above contains an `output` field. The `output` field must embeds a standard KubeVela application, consisting components, policies, and workflow.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+
+> Notice: This CUE file must have a `package main` header.
+
+In this example, we also define a namespace in `k8s-objects` component, and the name of it is set by parameters. If you want the created namespace is `my-namespace` you can execute the command like this:
```
If you want the created namespace to be `my-namespace`, you can run the following command:

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+> Please notice: The name of Application in template file will be replaced by the addon name in `metadata.yaml`. The addon application will always have a unified name in the format of `addon-<addon_name>`.
 
-#### CUE format resource
+Although you can define the entire content of the application in the template file, but this may cause the template file to become very large, so you can choose to split a part of some content into the `resources` directory, latter section will introduce the all details.
```
In case your template file is too large, you can split the entire content of the application into multiple files under the `resources` folder. The following section will discuss it in detail.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
-If you need to add a component to your application that needs to be rendered dynamically based on parameters when enabled, you can write a CUE format file as shown below:
+##### Auxiliary resources
+
+You also can define some auxiliary resources that only to applied to control-plane in `outputs` CUE block of `template.cue` file. These resources will applied to control-plane directly. For example:
```
You can also define some auxiliary resources in the `outputs` field of the `template.cue` file. These resources will only be applied to the control plane.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+
+##### Determine which clusters to be installed by parameters in `template.cue`
+
+If you want the resources in the addon to be installed not only in the control-plane, but also in various other managed-clusters, and installed clusters need to be specified by the user by setting the `clusters` parameter when enabling the addon. You can do this by adding a topology policy in `template.cue` as shown below:
```
If you want the resources in the addon to be installed not only in the control-plane, but also in managed clusters,  you can use the `topology` policy in your application as shown below. The `parameter.clusters` field will be filled when the addon is enabled with the `clusters` parameter specified.

wangyikewxgm committed:
fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

versin 1.5 i18n

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

fix comments of i18n

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
+          - cluster1
+```
+
+Since an empty `clusterLabelSelector` topology will choose all exist clusters as target clusters, you can just enable the addon without `clsuters` parameter to install it to all clusters:
````
without `clusters` parameters

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
+$ vela addon enable <addon-name>
+```
+
+The render result is :
````
The rendering result is:

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+
+### parameter.cue (Optional)
+
+In the example above, we used `parameter.namespace` parameter to set the name of namespace, and the `parameter.clusters` set the clusters that to be installed. We also need a parameter definition file (`parameter.cue`) to define what parameters the addon has. For examples:
```
we used -> we use

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
-It can cooperate with the resource file in CUE, here we have a CUE file `vela-apiserver.cue` :
+### `resources/` directory (Optional)
+
+Except writing every thing about an application in  one `temaplte.cue` or `template.yaml`, you also can split some resource description file to `resources` directory to avoid a very huge template file. File type in this dir can be YAML or CUE.
```
every thing -> everything
in  one -> in one
some resource description files to the `resources` directly
You can place either YAML files or CUE files in this directory.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+
+#### YAML resources
+
+The YAML type files must be Kubernetes objects, you can define many objects one by one in a file. It will be directly added to the application as a `K8s-object` type component during rendering. For example:
```
as a `k8s-objects` typed component

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
+- name: my-secret
+```
+
+In this example, we define a service account resource. The render result is :
````
define a ServiceAccount resource

Somefive commented on docs/platform-engineers/addon/intro.md:
````diff
@@ -100,175 +107,493 @@ metadata:
+            - name: my-secret
+```
+
+An example is [OCM](https://github.com/kubevela/catalog/tree/master/addons/ocm-hub-control-plane). All files in this addon is YAML.
````
is the OCM addon. All files included in this addon are all YAML typed.

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
-##### Use context render component
+#### Use metadata of context 
+
+In addition to dynamically rendering the application by parameters, you can also read fields defined in `metadata.yaml` for rendering. For example, you can define a `templatec.cue` file as follows:
```
you can define a `template.cue` file as follows:

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
 
-The image tag becomes the addon's version due to the `context.metadata.version` points to. The real example is [VelaUX](https://github.com/kubevela/catalog/blob/master/addons/velaux/resources/apiserver.cue).
-Other available fields please refer to [metadata](#metadata.yaml(Required)).
+The image tag becomes the addon's version due to the `context.metadata.version` points to. The real example is [VelaUX](https://github.com/kubevela/catalog/blob/master/addons/velaux/resources/apiserver.cue). Other available fields please refer to [metadata](#metadata.yaml(Required)).
```
Is this anchor working?

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+The image tag becomes the addon's version due to the `context.metadata.version` points to. The real example is [VelaUX](https://github.com/kubevela/catalog/blob/master/addons/velaux/resources/apiserver.cue). Other available fields please refer to [metadata](#metadata.yaml(Required)).
 
-UX/CLI renders all CUE files , `parameter.cue` and data defined in `metadata.yaml` in one context when the addon is enabled, resulting in a set of components that are appended to the application template.
+### `definitions/` directory(Optional)
```
directory (Optional)

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
-UX/CLI renders all CUE files , `parameter.cue` and data defined in `metadata.yaml` in one context when the addon is enabled, resulting in a set of components that are appended to the application template.
+### `definitions/` directory(Optional)
 
+The `definitions` directory is used to store `X-Definitions`, which can be a YAML file of ComponentDefinition, traitDefinitions or workflowStepDefinitions Kubernetes CR. It can also be a CUE type of  KubeVela [def](../../getting-started/definition), which will be rendered into the corresponding X-definitions CR when enabled and applied to the cluster.
```
TraitDefintions or WorkflowStepDefinitions

Somefive replied:
It can also be a CUE file that defines KubeVela Definitions

Somefive commented on docs/platform-engineers/addon/intro.md:
```diff
@@ -100,175 +107,493 @@ metadata:
+The `definitions` directory is used to store `X-Definitions`, which can be a YAML file of ComponentDefinition, traitDefinitions or workflowStepDefinitions Kubernetes CR. It can also be a CUE type of  KubeVela [def](../../getting-started/definition), which will be rendered into the corresponding X-definitions CR when enabled and applied to the cluster.
 
-#### YAML format resource
+> Please notice: These definitions only will be applied to control-plane
```
will only be applied to the control plane

wangyikewxgm committed:
small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
```diff
@@ -2,39 +2,49 @@
 ---
 
-一个 KubeVela 的插件就是包含了一系列文件的目录。 下图展示了在启用一个插件时，KubeVela 做了哪些事情, 主要包含以下三个步骤。
+一个 KubeVela 插件就是一个主要包含了以下三类文件的文件集合:
```
```suggestion
一个 KubeVela 插件就是一个主要包含了以下三类文件的集合:
```

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/intro.md:
````diff
@@ -72,211 +79,64 @@ invisible: false
-```shell
-vela addon enable velaux serviceAccountName="my-account"
-```
+#### 模块化能力定义文件 (definitions/ 目录) 
````
按钉钉说的，加上readme 那种结构性的内容，比如：

> `definitions` 文件包含了用户在安装该插件后能够看到的能力。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,85 @@
+title: 使用 YAML 描述插件应用
+---
+
+文档 [自定义插件](./intro) 介绍了插件的基本目录结构，并且介绍了插件中所需要安装的 Kubernetes operator 应当被定义在一个 KubeVela 应用（Application）中。本文档将会详细介绍如何使用 YAML 格式的应用描述文件来定义这个应用。
```
好像是个病句。

并且介绍了插件中所需要安装的 Kubernetes operator 如何被定义在一个 KubeVela 应用（Application）中

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,85 @@
+
+在这个例子中，我们定义的应用基础框架中包含了一个` k8s-objects` 类型的组件，组件中只包含了一个 Kubernetes namespace。
+
+> 需要注意的是，即使你在应用模版文件中设置了应用的名称，该设置也不会生效，在启用时应用会统一以 addon-{addonName} 的格式自动命名。
```
那上面那个例子是不是可以把 name 和 namespace 删掉？不然误导了

wonderflow replied:
这个yaml 的模板有很多限制和自动填充的过程是不是要解释？

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-yaml.md:
````diff
@@ -0,0 +1,85 @@
+              secrets:
+                - name: my-secret
+```
+
````
加一句话，表达一下其他的 CRD、Controller也是同样的方式。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,411 @@
+```
+
+请注意这个应用中的 `spec.components[0].properties.objects[0]` 中所所定义的 Kubernetes namespace 的名称是由 `parameter.namespace` 所决定的。这代表它的名字在启用插件时会被 `namespace` 启动参数所动态渲染。
+
````
这里可以引导一下用户去了解 CUE 的语法，我们有个文档是干这个介绍的。

wangyikewxgm replied:
我在 348 行提到了这个文档，意思挪到上面来吗？

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,411 @@
+  namespace?: foo-namespace | string
+}
+```
+
````
这个参数定义的效果可以在这里简单介绍一下。 vela addon --xxx

wonderflow replied:
我看下面介绍了，你可以在这里提一下“下面会介绍”，顺承下去。不然用户会觉得不理解

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,411 @@
+          - local
+          - cluster1
+```
+
````
这话感觉都没说完，至少补一句。

Addon 安装后 KubeVela 应用执行过程会根据 topology 策略将这个插件包含的组件部署到 `local` 和 `cluster1` 两个集群。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,411 @@
+      properties:
+        clusterLabelSelector: { }
+```
+
````
这里也是没说完，topology 是个空的花括号代表什么意思呢？

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,411 @@
+
+上面就完整的介绍了如何用 CUE 编写 `template.cue`, `parameter.cue` 以及 `resources/` 目录下的 CUE 资源文件，在启用插件时这些文件会连同 `metadata.yaml` 中记录的插件元信息在同一个上下文中渲染得到结果并下发。
+
+如果你对 CUE 语言还不了解，可以通过 [CUE 基础入门文档](../cue/basic) 了解 CUE 的具体语法。
```
这个应该放前面的，能看到这里的读者100% 都不需要读这个文档了。因为不懂 CUE 的人在看到第一个CUE的代码块就放弃阅读了。

wonderflow replied:
你如果想表达高级用法，就说，你还可以在 Addon 中使用其他 CUE 语言的高级特性，比如你可以在本地使用 `cue eval *.cue resources/*.cue -e output -d yaml` 命令查看资源渲染的效果。

wonderflow commented on i18n/zh/docusaurus-plugin-content-docs/current/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,411 @@
+```shell
+$ vela addon enable fluxcd onlyHelmComponents=true
+```
+
````
onlyHelmComponents=true

本质上是上面参数的那个能力吧，这个应该说一句。不要给完一个命令就结束。

wangyikewxgm committed:
fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
fix english

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
refine the struct

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

Somefive commented on docs/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,79 @@
+title: YAML application description file
+---
+
+[Make Your Own Addon] introduces the basic structure of an addon, and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. This doc will discuss all the details of writing the application description file with YAML.
```
[Make Your Own Addon] lacks link?

Somefive commented on docs/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,79 @@
+
+In case your template file is too large, you can split the entire content of the application into multiple files under the `resources/` folder.
+
+YAML file in`resources/` folder must be Kubernetes objects, you can define many objects one by one in a file. These objects will be directly added to the application as a `K8s-object` typed component during rendering. An example as follows:
```
K8s-object -> k8s-objects

Somefive commented on docs/platform-engineers/addon/addon-yaml.md:
````diff
@@ -0,0 +1,79 @@
+            - name: my-secret
+```
+
+We just use namespace and serviceAccount resource as example here, other resources of an operator can also be defined in KubeVela application in the same way.
````
resource -> resources
as example -> as an example

Somefive commented on docs/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,79 @@
+
+## Example
+
+An example is [OCM](https://github.com/kubevela/catalog/tree/master/addons/ocm-hub-control-plane). All files included in this addon are all YAML typed.
```
YAML typed -> YAML coded ?

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+title: CUE application description file
+---
+
+[Document](./intro) introduces the basic structure of an addon and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. Document [YAML application description file](./addon-yaml) explains the way using YAML define the application. If you use CUE to write application description file the addon will be able to have these abilities:
```
Document -> Previous tutorial ?

Somefive replied:
the way of using YAML to define the addon application

Somefive replied:
use CUE to write application files, the addon will have the following capabilities

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+[Document](./intro) introduces the basic structure of an addon and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. Document [YAML application description file](./addon-yaml) explains the way using YAML define the application. If you use CUE to write application description file the addon will be able to have these abilities:
+
+* Utilize the flexible and concise syntax of the CUE language, rich built-in functions and its parameter verification capabilities, to render and deploy the application and auxiliary resources with parameters and metadata of addon.
```
parameters and metadata of the addon

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+[Document](./intro) introduces the basic structure of an addon and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. Document [YAML application description file](./addon-yaml) explains the way using YAML define the application. If you use CUE to write application description file the addon will be able to have these abilities:
+
+* Utilize the flexible and concise syntax of the CUE language, rich built-in functions and its parameter verification capabilities, to render and deploy the application and auxiliary resources with parameters and metadata of addon.
+* An addon may contain multiple Definitions and CRD Operators, they can be selectively installed according to parameters of addon.
```
Operators. They can be

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+* Utilize the flexible and concise syntax of the CUE language, rich built-in functions and its parameter verification capabilities, to render and deploy the application and auxiliary resources with parameters and metadata of addon.
+* An addon may contain multiple Definitions and CRD Operators, they can be selectively installed according to parameters of addon.
+
+This doc will introduce how to Write the application description file using CUE.
```
how to write application description files using CUE
/
how to define addon application by writing CUE files

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+}
+```
+
+In this example, the name of the namespace defined in `spec.components[0].properties.objects[0]` in this application is determined by `parameter.namespace`, which means that its name will be dynamically rendered by the `namespace` parameter when the addon is enabled. If you want the created namespace to be my-namespace, you can run the following command:
````
to be `my-namespace`

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+## Parameter definition file (parameter.cue) 
+
+In the example above, we use the parameter `namespace`  to set the name of namespace resource. Actually, We also need a parameter definition file (`parameter.cue`) to declare what parameters this addon has. For examples:
```
Actually, we also need

Somefive replied:
For example,

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+When enabling the addon, you can set the parameters declared in the parameter definition file by appending the parameters to the command, as follows:
+
+```shell
+$ vela addon enable <addon-Name> <parameter-name-1=value> <parameter-name-1=value>
````
<parameter-name-2=value>

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+}
+```
+
+Then we can reference the two CUE block in `template.cue` :
````
two CUE blocks

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+              name: my-namespace
+```
+
+> Please notice: Only those CUE files with header `package main` can be reference by `template.cue`, this can be used to help you filter CUE files that you don't want to set into the rendering context.
````
you don't want to use in the rendering context

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+        clusterLabelSelector: {}
+```
+
+Since an empty (`{}`) `clusterLabelSelector` topology will choose all exist clusters as target, so the components in application will be dispatched to all clusters including control-plane and managed-cluster.
````
as target, the components in the application will be dispatched to all clustering including both the control plane and the managed clusters.

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+### Auxiliary resources
+
+You can also define some auxiliary resources in the outputs field of the `template.cue` file. These resources will only be applied to the control plane.
```
in the `outputs` field

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+        image: "oamdev/vela-apiserver:v1.2.4"
+```
+
+The image tag becomes the addon's version due to the `context.metadata.version` points to. The real example is [VelaUX](https://github.com/kubevela/catalog/blob/master/addons/velaux/resources/apiserver.cue). Other available fields of metadata please refer to [metadata](./intro).
````
The image tag becomes the addon's version which the `context.metadata.version` field points to.

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+### Binding the Definition to a component
+
+If you want to bind a Definition to a component in the application, to achieve dynamically enable the ability of one X-definition, you can do it by setting `addon.oam.dev/bind-component` annotation on the definition.
```
to dynamically enable the ability of one X-definition

Somefive commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+//...
+```
+
+If you enable this addon by following command:
````
by the following command

wangyikewxgm committed:
fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wangyikewxgm committed:
small fix

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+---
+title: CUE application description file
```
```suggestion
title: CUE based Addon Application
```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+This doc will introduce how to define addon application by writing CUE files.
+
+Application description files contain two parts: application template files and resource files (files in the `resources/` folder).
```
```suggestion
Application description files contain two parts: application template files and resource files (they're in the `resources/` folder).
```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+## Application template file (template.cue)
+
+The most important part in the application template is `output` field, which must embed a KubeVela application as follows:
```
```suggestion
The most important part in the application template is `output` field, which must place a KubeVela application as follows:
```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+}
+```
+
+In this example, the name of the namespace defined in `spec.components[0].properties.objects[0]` in this application is determined by `parameter.namespace`, which means that its name will be dynamically rendered by the `namespace` parameter when the addon is enabled. If you want the created namespace to be `my-namespace`, you can run the following command:
````
```suggestion
In this example, the name of the namespace defined in `spec.components[0].properties.objects[0]` of this application is determined by `parameter.namespace`, which means that its name will be dynamically rendered by the `namespace` parameter when the addon is enabled. If you want the created namespace to be `my-namespace`, you can run the following command:
```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+kind: Application
+metadata:
+  name: example
+  namespace: vela-system
```
```suggestion

```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+## Parameter definition file (parameter.cue) 
+
+In the example above, we use the parameter `namespace`  to set the name of namespace resource. Actually, we also need a parameter definition file (`parameter.cue`) to declare what parameters this addon has. For example,
```
```suggestion
In the example above, we use the parameter `namespace`  to set the name of Kubernetes `namespace` resource. Actually, we also need a parameter definition file (`parameter.cue`) to declare what parameters this addon has. For example,
```

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
````diff
@@ -0,0 +1,407 @@
+When enabling the addon, you can set the parameters declared in the parameter definition file by appending the parameters to the command, as follows:
+
+```shell
+$ vela addon enable <addon-Name> <parameter-name-1=value> <parameter-name-2=value>
````
```suggestion
vela addon enable <addon-Name> <parameter-name-1=value> <parameter-name-2=value>
```

如果只是一行命令就不加 $ 了，方便用户copy。

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+## Resource files (CUE files under `resources/` folder)
+
+In case your template file is too large, you can split the entire content of the application into multiple files under the `resources/` folder.
```
这里的点显得太小了，这里要从 CUE 的语法支持说.

KubeVela has supported CUE package in addon, which means you can define any CUE files containing data or schema inside the `resources/` folder and reference them in application CUE file as a while package.

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+In case your template file is too large, you can split the entire content of the application into multiple files under the `resources/` folder.
+
+Unlike the YAML resource file, the CUE resource file defines the CUE block that can be referenced by `template.cue`
```
这里跟上面结合着说

wonderflow commented on docs/platform-engineers/addon/addon-cue.md:
```diff
@@ -0,0 +1,407 @@
+
+> Please notice: Only those CUE files with header `package main` can be reference by `template.cue`, this can be used to help you filter CUE files that you don't want to use in the rendering context.
+
+We just use namespace as example here, other resources of an operator can also be defined in KubeVela application in the same way.
```
```suggestion
We just use namespace as example here, other resources of an operator can also be defined in KubeVela application in the same way. This also gives your addon reusablity and validation capability powered by CUE.
```

Make your vision bigger 😄

wonderflow commented on docs/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,79 @@
+title: YAML application description file
+---
+
+[Make Your Own Addon](./intro) introduces the basic structure of an addon, and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. This doc will discuss all the details of writing the application description file with YAML.
```
```suggestion
[Make Your Own Addon](./intro) introduces the basic structure of an addon, and illustrate that any Kubernetes operator to be installed of an addon should be defined in a KubeVela application. This doc will guide you with all the details of writing the application description file with YAML.
```

wonderflow commented on docs/platform-engineers/addon/addon-yaml.md:
```diff
@@ -0,0 +1,79 @@
+
+## Application template file (template.yaml)
+
+The YAML typed application template file must define a KubeVela application that can contain components, policies or workflow. A simple example is as follows: 
```
```suggestion
The YAML typed application template file only accepts KubeVela application. A simple example is as follows: 
```

想说 Application包含什么，就加个 reference 链接到 Application 概念。更好一些。

wonderflow reviewed (approved):
LGTM after nits

wangyikewxgm committed:
fix comments

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

fix typo

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

wonderflow reviewed (approved):
great work
